src/client/Client.cc

```diff
@@ -7945,18 +7945,9 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
     ldout(cct,10) << "changing size to " << stx->stx_size << dendl;
     if (in->caps_issued_mask(CEPH_CAP_FILE_EXCL) &&
         !(mask & CEPH_SETATTR_KILL_SGUID) &&
-        stx->stx_size >= in->size) {
-      if (stx->stx_size > in->size) {
-        in->size = in->reported_size = stx->stx_size;
-        in->cap_dirtier_uid = perms.uid();
-        in->cap_dirtier_gid = perms.gid();
-        in->mark_caps_dirty(CEPH_CAP_FILE_EXCL);
-        mask &= ~(CEPH_SETATTR_SIZE);
-        mask |= CEPH_SETATTR_MTIME;
-      } else {
-        // ignore it when size doesn't change
-        mask &= ~(CEPH_SETATTR_SIZE);
-      }
+        stx->stx_size == in->size) {
+      // ignore it when size doesn't change
+      mask &= ~(CEPH_SETATTR_SIZE);
```

Comment:
So this is actually a pretty impactful change in terms of performance, right? This is one of a very limited number of places we actually make use of FILE_EXCL!

If I understand right, the concern is that a client with Fx doesn't update the rstats for that file very often, and this is breaking quota tests. But I'm not sure dropping this behavior completely is the right way to resolve that. Can you discuss @lxbsz?

Reply:
Although this is really just for doing truncates? Why do we even want to cut that short then — does it require a big set of cap swaps?

Reply:
Yeah, this will affect the perf anyway. 

As you know we need to check all the parent quota realms when checking the quota.

The problem is we need to update all the parent quota realms once the file size changed. Before I figured out one way to improve this, which is when the size changes are cached in client side we can just update the `rstat` for all the parent quota realms locally. 

The above method will work for single client case, but for the multiple clients case it seems won't. For example:

```
                                  dirA (quota_realmA)
                                /         \
                         dirB1    dirB2 (quota_realmB2)
                            /                \
   dirC1(quota_realmC1)    dirC2
                        /                         \
                    fileD1                   fileD2 
```

If the `fileD1` is opened in `client1` and `fileD2` is opened in `client1`. If in `client1` we change the size for `fileD1` and cache it locally. We need to notify this to `client2` or when the `client2` try to write to `fileD2` the size quota check in `quota_realmA` will be incorrect.

Maybe to improve this we still can cache the size truncate here, then the `setattr()` could just return success immediately just after updating all the parent quota realms and then flush the dirty caps to MDSs immediately too.  

But there possibly be a quota mismatch issue between clients for a short time ?

Will this make sense to you @gregsfortytwo ?

Reply:
From reading the code these days and when debugging other issues. It seems cephfs won't guarantee that the rstat will be applied and notified to the clients immediately. 

Such as the `mds_dirstat_min_interval` option, which the default value is `1`, will avoid updating the dirstat too often and will defer updating it. That means when the dirty `Fw` caps is flushed to MDS it may skip updating the parent dir's rstat. Then the clients couldn't get the notification for the changes.

So in this case I will fix this in the test code instead.

Reply:
My concern is that we already know the quotas aren't strict, and this is making a tradeoff against other functionality to make quotas stricter. And as you've said, rstats are already not a precisely-timed update mechanism.

So I think we just need to accept that, and make sure the tests don't break on it. Unless we have a specific reason that keeps in mind the limits of the quota support we have, and still suggests we should be strengthening them like this.